python/vowpalwabbit/DFtoVWtoTensorboard.py

```diff
@@ -0,0 +1,208 @@
+										This is the difference between current iteration's sum_loss and previous iteration's sum_loss which is used to calculate since_last
+		weighted_examples_since_last : float
+										This is the difference between current iteration's weighted_examples and previous iteration's weighted_examples which is used to calculate since_last
+
+		Returns
+		-------
+
+		since_last      : float
+							since_last calculated using sum_loss_since_last and weighted_examples_since_last or if ZeroDivisionError is encountered then 0. is returned 
+		""" 
+		try:
+			return sum_loss_since_last / weighted_examples_since_last
+
+		except ZeroDivisionError:
+			return 0.
+
+
+	#------------------------------------------------------------------------------------------------		
+	def _print_metrics(self, average_loss, since_last, label, prediction, num_features):
+		"""This method prints metrics
+
+		Parameters
+		----------
+
+		average_loss : float
+						This is the average_loss value
+		since_last   : float
+						This is the since_last value
+		label        : integer
+						This is the actual label of current example
+		prediction   : integer
+						This is the predicted label of current example
+		num_features : integer
+						This is the num_features value
+
+		Returns
+		-------
+
+		None
+		"""
+		print( 'average_loss:{:.6f}'.format(average_loss) , end='\t')
+		print('since_last:{:.6f}'.format(since_last), end='\t')
+		print('label:', label, end='\t')    
+		print('prediction:', prediction, end='\t')    
+		print('num_features:', num_features) 
+
+
+	#------------------------------------------------------------------------------------------------		
+	def get_label(self, example, label_type):
+		"""Similar to pyvw.get_prediction implementation this method gets the label for current example
+
+		Parameters
+		----------
+
+		example    : example object
+						Methods of this example object are used to get the appropiate label
+		label_type : integer
+						This is the integer for representing a specific type of label in vw
+						- 0: lDEFAULT
+						- 1: lBINARY
+						- 2: lMULTICLASS
+						- 3: lCOST_SENSITIVE
+						- 4: lCONTEXTUAL_BANDIT
+						- 5: lMAX
+						- 6: lCONDITIONAL_CONTEXTUAL_BANDIT
+						- 7: lSLATES
+						- 8: lCONTINUOUS
+
+		Returns
+		-------
+
+		label  : integer
+					Actual label of the current example object
+		"""
+		switch_label_type = {
+			pyvw.vw.lDefault: None,
+			pyvw.vw.lBinary: example.get_simplelabel_label,
+			pyvw.vw.lMulticlass: example.get_multiclass_label,
+			pyvw.vw.lCostSensitive: example.get_costsensitive_class,
+			pyvw.vw.lContextualBandit: example.get_cbandits_class,
+	#         pyvw.vw.lConditionalContextualBandit: None,
+	#         pyvw.vw.lSlates: None,
+	#         pyvw.vw.lContinuous: None
+		}
+		return switch_label_type[label_type]()	
+
+
+	#------------------------------------------------------------------------------------------------		
+	def fit(self, df, tensorboard=True):
```

Comment:
A lot of this method shares code with `vw.learn` - consider unifying them (e.g. pull in the Tensorboard code into `.learn`, and either add a path to take in a DF in the method, or create a `learn_all` or `fit` method on the vw class.

cc @etiennekintzler, @jackgerrits - do you have an opinion on whether we should reuse the `.learn` method vs. a `learn_all`/`fit` method to learn over the entire DFtoVW dataset.

@

Reply:
@lokitoth In my opinion the whole class is a bit overkill and embedding the fitting process inside a  `fit` method is too restrictive.

I am not super familiar with tensorboard but the writer is usually used inside the fitting loop, for example see the following pytorch reference: https://pytorch.org/tutorials/intermediate/tensorboard_tutorial.html . Also it is up to the user to include what he wants in the tensorboard, which is not allowed here.

Instead of this class, I think what could be great is a notebook that uses Tensorboard in the same fashion as the pytorch tensorboard tutorial. What do you think of this idea @speedhunter001 @lokitoth ?

Reply:
Another thing that could be done is to build a helper class that parse the standard vowpal wabbit output (see below) and write it to files that can be rendered with tensorboard.

```
Num weight bits = 18
learning rate = 0.5
initial_t = 0
power_t = 0.5
using no cache
Reading datafile = test.csv
num sources = 1
Enabled reductions: gd, scorer
average  since         example        example  current  current  current
loss     last          counter         weight    label  predict features
1.000000 1.000000            1            1.0   1.0000   0.0000       15
0.500000 0.000000            2            2.0   1.0000   1.0000       15
0.504930 0.509860            4            4.0   1.0000   0.5378       15
0.323170 0.141410            8            8.0   1.0000   1.0000       15
....
```

It could be useful to anyone who want to have graphical summary of the performances/parameters and would not require a tight coupling with the user code.

Reply:
@etiennekintzler To get familiar with Tensorboard, I modified the following tutuorial on **Content Personalization Simulation**  https://vowpalwabbit.org/tutorials/cb_simulation.html  to see the ctr (click through rate) plots on Tensorboard.  By seeing such an example, the user may get an idea to log any scalar metrics for Tensorboard. So what do you suggest on this that should I include this as an example in PR too?   

The current class DFtoVWtoTensorboard **logs the following standard vowpal wabbit output metrics i.e average_loss, since_last, current_features** which one can see in Tensorboard, currently Tensorboard does not have support for bar plots so we thought that logging **current label** and **current_predict** would not be suitable. What is your suggestion on this?

Reply:
@lokitoth I would try to understand the code of `vw.learn` to consider making the changes

Reply:
> The current class DFtoVWtoTensorboard logs the following standard vowpal wabbit output metrics i.e average loss, since last, current features which one can see in Tensorboard, this currently provides the user to see few of the standard vowpal wabbit output metrics, currently Tensorboard does not have support for bar plots so we thought that logging current label and current predict would not be suitable. What is your suggestion on this?

**Your class computes metrics that are already available in the default vowpal wabbit output** (the output in my last message). 

My suggestion was to **just parse this VW output** and write the parsed result to tensorboard. Below is a notebook that essentially do this: https://github.com/etiennekintzler/vowpal_wabbit/blob/tensorboard_example/python/examples/tensorboard_example/tensorboard_vw.ipynb

Also, this way to do is a bit more generic as you just need the output file and you don't need to redo the whole fitting process inside the class.

Below is the plot of tensorboard I obtained : 

![image](https://user-images.githubusercontent.com/17943873/120717446-d968eb80-c4c7-11eb-8222-27cab056a504.png)

Reply:
@etiennekintzler Rather than parsing the VW output, it seems like it would be good to enable a callback into python directly exposing these metrics (and short-circuiting the printout entirely), or having them be available in a dictionary (like the extra metrics that @lalo added in #3022). That would be a longer-term goal of @speedhunter001's efforts here this summer.

Reply:
If the metrics are directly available from python's VW model it's indeed more straightforward to use it directly to write into tensorboard files

Good luck and thanks to @speedhunter001 for taking care of this :)